src/main/scala/mesosphere/marathon/api/akkahttp/v2/AppsController.scala

```diff
@@ -112,44 +153,331 @@ class AppsController(
+    val version = clock.now()
+
+    (forceParameter &
+      extractClientIP &
+      extractUri &
+      entity(as(appUpdateUnmarshaller(appId, partialUpdate)))) { (force, remoteAddr, requestUri, appUpdate) =>
+        /* Note - this function throws exceptions and handles authorization synchronously. We need to catch and map these
+       * exceptions to the appropriate rejections */
+        val fn = updateOrCreate(
+          appId, _: Option[AppDefinition], appUpdate, partialUpdate, allowCreation, clock.now(), marathonSchedulerService)
+
+        onSuccessLegacy(Some(appId))(groupManager.updateApp(appId, fn, version, force)).apply { plan =>
+          plan.target.app(appId).foreach { appDef =>
+            eventBus.publish(ApiPostEvent(remoteAddr.toString, requestUri.toString, appDef))
+          }
+
+          completeWithDeploymentForApp(appId, plan)
+        }
+      }
+  }
+
+  private[this] def updateMultiple(partialUpdate: Boolean, allowCreation: Boolean)(implicit identity: Identity) = {
+    val version = clock.now()
+
+    (forceParameter & entity(as(appUpdatesUnmarshaller(partialUpdate)))) { (force, appUpdates) =>
+
+      def updateGroup(rootGroup: RootGroup): RootGroup = appUpdates.foldLeft(rootGroup) { (group, update) =>
+        update.id.map(PathId(_)) match {
+          case Some(id) => group.updateApp(id, AppHelpers.updateOrCreate(id, _, update, partialUpdate, allowCreation, clock.now(), marathonSchedulerService), version)
+          case None => group
+        }
+      }
+
+      onSuccessLegacy(None)(groupManager.updateRoot(PathId.empty, updateGroup, version, force)).apply { plan =>
+        complete((StatusCodes.OK, List(Headers.`Marathon-Deployment-Id`(plan.id)), Messages.DeploymentResult(plan)))
+      }
+    }
+  }
+
+  private def completeWithDeploymentForApp(appId: PathId, plan: DeploymentPlan) =
+    plan.original.app(appId) match {
+      case Some(_) =>
+        complete((StatusCodes.OK, List(Headers.`Marathon-Deployment-Id`(plan.id)), Messages.DeploymentResult(plan)))
+      case None =>
+        complete((StatusCodes.Created, List(Location(Uri(appId.toString)), Headers.`Marathon-Deployment-Id`(plan.id)), Messages.DeploymentResult(plan)))
+    }
+
+  /**
+    * It'd be neat if we didn't need this. Would take some heavy-ish refactoring to get all of the update functions to
+    * take an either.
+    */
+  private def onSuccessLegacy[T](maybeAppId: Option[PathId])(f: => Future[T])(implicit identity: Identity): Directive1[T] = onComplete({
+    try { f }
+    catch {
+      case NonFatal(ex) =>
+        Future.failed(ex)
+    }
+  }).flatMap {
+    case Success(t) =>
+      provide(t)
+    case Failure(ValidationFailedException(_, failure)) =>
+      reject(EntityMarshallers.ValidationFailed(failure))
+    case Failure(AccessDeniedException(msg)) =>
+      reject(AuthDirectives.NotAuthorized(HttpPluginFacade.response(authorizer.handleNotAuthorized(identity, _))))
```

Comment:
Same here. Do we not check authorization *before* we get into the controller?

Reply:
It might happen, but I'll refactor it.